packages/apollo-language-server/src/schema/providers/engine.ts

```diff
@@ -15,29 +21,54 @@ import {
 
 export class EngineSchemaProvider implements GraphQLSchemaProvider {
   private schema?: GraphQLSchema;
-  private client?: ApolloEngineClient;
+  private engineClient?: ApolloEngineClient;
 
-  constructor(private config: ClientConfig) {}
+  constructor(private config: ClientConfig | ServiceConfig) {}
   async resolveSchema(override: SchemaResolveConfig) {
     if (this.schema && (!override || !override.force)) return this.schema;
-    const { engine, client } = this.config;
+    const { engine, client, service } = this.config;
 
-    if (typeof client.service !== "string") {
-      throw new Error(
-        `Service name not found for client, found ${client.service}`
-      );
+    let serviceName;
+
+    if (isClientConfig(this.config)) {

```

Comment:
This one is a bit tricky. The isClientConfig check doesn't explicitly check for `config.client`. `client` is a required field for `ClientConfig`, but not `ServiceConfig` and since `this.config` is a union between the two, theoretically, client _could_ be null

Reply:
Interesting, I see what you mean. I wonder if the `isClient` and `isService` fields are extraneous. Should be just checking for presence of client or service field I think.

Out of scope though 😛

Reply:
I think `isClientConfig` and `isServiceConfig` guarantee `client` or `service` are non-null, because they acts as type guards (`config is ClientConfig` or `config is ServiceConfig` respectively). So if you move the destructuring within the conditionals, you shouldn't need `!s`.